src/communication/RpcClient.cpp

```diff
@@ -10,3 +10,371 @@
+	    all_expired;
+
+	c.erase(
+	    std::remove_if(c.begin(), c.end(),
+	                   [instance_id, &all_expired](const PendingRequest& p) {
+		                   if (instance_id == p.instance_id) {
+			                   all_expired.push_back(p.expire);
+			                   return true;
+		                   }
+		                   return false;
+	                   }),
+	    c.end());
+
+	// TODO - is there a better way to shrink the internal container?
+	// Maybe instead we should enforce a capacity limit
+	constexpr size_t capacity_shrink_threshold = 16;
+	if ((c.capacity() > capacity_shrink_threshold) &&
+	    (c.size() < c.capacity() / 2)) {
+		c.shrink_to_fit();
+	}
+
+	return all_expired;
+}
+
+// Exposing non-const version so the listen handle can be moved out
+PendingRequest& ScrubablePendingQueue::top() { return c.front(); }
+
+ExpireWorker::ExpireWorker() {
+	worker_ = std::thread([this]() { doWork(); });
+}
+
+ExpireWorker::~ExpireWorker() {
+	stop_ = true;
+	{
+		std::lock_guard lock(pending_mtx_);
+		wake_worker_.notify_one();
+	}
+	worker_.join();
+}
+
+void ExpireWorker::enqueue(PendingRequest&& pending) {
+	std::lock_guard lock(pending_mtx_);
+	pending_.emplace(std::move(pending));
+	wake_worker_.notify_one();
+}
+
+void ExpireWorker::scrub(size_t instance_id) {
+	std::vector<std::function<void(communication::RpcClient::Status)>>
+	    all_expired;
+	{
+		std::lock_guard lock(pending_mtx_);
+		all_expired = pending_.scrub(instance_id);
+		wake_worker_.notify_one();
+	}
+
+	static const v1::UStatus cancel_reason = []() {
+		v1::UStatus reason;
+		reason.set_code(v1::UCode::CANCELLED);
+		reason.set_message("RpcClient for this request was discarded");
+		return reason;
+	}();
+
+	for (auto& expire : all_expired) {
+		expire(cancel_reason);
+	}
+}
+
+void ExpireWorker::doWork() {
+	while (!stop_) {
+		const auto now = std::chrono::steady_clock::now();
+		std::optional<decltype(PendingRequest::expire)> maybe_expire;
+
+		{
+			transport::UTransport::ListenHandle expired_handle;
+			std::lock_guard lock(pending_mtx_);
+			if (!pending_.empty()) {
+				const auto when_expire = pending_.top().when_expire;
+				if (when_expire <= now) {
+					maybe_expire = pending_.top().expire;
+					expired_handle =
+					    std::move(pending_.top().response_listener);
+					pending_.pop();
+				}
+			}
+		}
+
+		if (maybe_expire) {
+			auto& expire = *maybe_expire;
+
+			static const v1::UStatus expire_reason = []() {
+				v1::UStatus reason;
+				reason.set_code(v1::UCode::DEADLINE_EXCEEDED);
+				reason.set_message("Request expired before response received");
+				return reason;
+			}();
+
+			expire(expire_reason);
+
+		} else {
+			std::unique_lock lock(pending_mtx_);
+			if (pending_.empty()) {
+				wake_worker_.wait(
+				    lock, [this]() { return stop_ || !pending_.empty(); });
+			} else {
+				auto wake_when = pending_.top().when_expire;
+				wake_worker_.wait_until(lock, wake_when, [this, &wake_when]() {
```

Comment:
The pending_ list can have elements who's responses have been met, yet stay in this list until they "expire" and then attempt (but fail due to the call_once flag) to call the callback.  The issue is that they are in the list the entire time of their TTL even if the RPC request completes "instantly".

Right before we "wait_until", we _could_ decide to clean those items out of the list if the wait_until time is far enough in the future.

Reply:
I think what I'm going to have to do to address this issue is

1. Add the request ID to the `PendingRequest` struct
2. Add a second `scrub()` method to `ScrubablePendingQueue` that takes a UUID to scrub a single request
3. Add a second `scrub()` method to `ExpireWorker` that takes a UUID and puts it in a scrub queue for deferred scrubbing (and then wakes the worker thread)
4. Modify the callback wrapper such that it calls that `scrub()` method on the `ExpireService` for the request ID inside of the `call_once` section and after the message has been passed to the caller.

Testing this is also going to be somewhat challenging. There isn't a direct way to observe if the request is still in the pending queue. I think I would need to add a protected `bool pending(v1::Uuid)` method to `RpcClient` (along with the supporting methods in the expiration service infrastructure). Then the test could inherit from RpcClient and use that `pending()` method to check.

I'm a little nervous about the scope of that change and the possible interactions this close to a release. @billpittman - how bad do you think this issue is, and could we possibly take it as an improvement in alpha.3?

Reply:
I've preemptively filed #221 to address this as a follow-up in alpha.3. Let me know what you think - can we cut the alpha.2 release with this issue or is it a critical bug?

Reply:
I agree this could have several ripples and isn't really worth holding the release.  Assuming the issue you filed doesn't get downgraded to the point of obscurity, I'm fine with waiting until the next open window to fix this.

Reply:
I'd say this conversation is resolved.

Reply:
I think you'll find I don't like letting issues like this fade away :grin: